src/store_client.cc

```diff
@@ -153,59 +155,58 @@ storeClientListAdd(StoreEntry * e, void *data)
     return sc;
 }
 
-void
-store_client::callback(ssize_t sz, bool error)
+static void
+CallbackClientSide(store_client *sc)
 {
-    size_t bSz = 0;
+    sc->callbackClientSide();
+}
 
-    if (sz >= 0 && !error)
-        bSz = sz;
+void
+store_client::callbackClientSide()
+{
+    assert(_callback.pending());
 
-    StoreIOBuffer result(bSz, 0,copyInto.data);
+    clientSideCaller = nullptr;
+    StoreIOBuffer result(copiedSize, copyInto.offset, copyInto.data);
```

Comment:
An alternative approach would be to avoid creating the temporary variable and use copyInto directly. This, in turn, would entail storing copiedSize in copyInto, as a separate field (which does not worth it, IMO) or overwriting copyInto.length with copiedSize after a successful read from Store. The latter option does not work also, because store_client::scheduleMemRead() may be called two times before proceeding with the client side callback. This happens after the response header is written (the first call) and the initial body bytes are written (the second call). The second call reads both header and initial body bytes (thus overwriting the already stored in copyInto header bytes). So if we, for example, save the header size in copyInto.length after the first call, the second call will be useless: we simply overwrite header bytes with header bytes. BTW, while testing, I noticed that the pointless 'overwriting' of header bytes happens when we make a range request.

Reply:
>  because store_client::scheduleMemRead() may be called two times before proceeding with the client-side callback.

This did not happen in master, because of StoreEntry::invokeHandlers() check:
```
if (!sc->_callback.pending())
    continue;
```
We can adjust this check in the PR and suppress these problematic storeClientCopy2() calls, overwriting the buffer:
```
if (!sc->_callback.pending() || clientSideCaller)
    continue;
```
In other words, once we have scheduled the initial clientSideCaller (upon receiving Http headers), we do not allow more notifications from Store (via invokeHandlers()) and let client-side code sending these headers and schedule clientSideCaller again.

Reply:
If master never touches copyInto twice in a row (i.e. without a notification in-between), in any contexts, then I am not against establishing/restoring similar protections (also in all contexts). If master could touch copyInto twice in some cases, then I would be worried that we may be breaking some unusual cases.

> upon receiving Http headers

I know that you are looking at some HTTP test cases, but Store is used for so much more that a simple HTTP transaction itself does not convince me that we can bypass some code paths under some conditions and everything would be fine (because it looks fine in a test case).

> Slightly paraphrased: Once we have scheduled the initial clientSideCaller, we block notifications from Store. The client-side code sends these headers and then schedules clientSideCaller again.

Is that what essentially happens in master?